// airplane/internal/display/table.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.filterRowsByColumns = exports.columnsFromRows = exports.table = void 0;
const runtime_1 = require("../runtime");
/**
 * Appends a display to the run that renders a table.
 *
 * Each row should be an object mapping header slugs to values. Columns that are not
 * specified will default to `null`. The selection, ordering, and naming of columns
 * can be customized via `opts.columns`.
 * @example
 *   await table([
 *     { "name": "Lilo", age: 6 },
 *     { "name": "Stitch" },
 *   ])
 * @param rows The list of rows to render in the table
 * @param opts Additional configuration options
 */
const table = async (rows, opts = {}) => {
    let cfg;
    if (opts.columns) {
        // Convert string[] -> DisplayTableColumnConfig[], if necessary
        const columns = opts.columns.length > 0 && typeof opts.columns[0] === "string"
            ? opts.columns.map((c) => ({ slug: c }))
            : opts.columns;
        cfg = {
            kind: "table",
            columns,
            rows: (0, exports.filterRowsByColumns)(rows, columns),
        };
    }
    else {
        cfg = {
            kind: "table",
            columns: (0, exports.columnsFromRows)(rows),
            rows,
        };
    }
    await (0, runtime_1.getRuntime)().display(cfg);
};
exports.table = table;
const columnsFromRows = (rows) => {
    const columns = [];
    const knownColumns = new Set();
    for (const row of rows) {
        for (const key of Object.keys(row)) {
            if (!knownColumns.has(key)) {
                columns.push({ slug: key });
                knownColumns.add(key);
            }
        }
    }
    return columns;
};
exports.columnsFromRows = columnsFromRows;
const filterRowsByColumns = (rows, columns) => {
    const knownColumns = new Set(columns.map((c) => c.slug));
    if (knownColumns.size !== columns.length) {
        throw new Error(`Expected column slugs to be unique`);
    }
    if (columns.some((c) => !c.slug)) {
        throw new Error(`Column slugs cannot be empty`);
    }
    const isKnownKey = (key) => {
        return knownColumns.has(key);
    };
    const filteredRows = [];
    for (const row of rows) {
        const keys = Object.keys(row);
        const knownKeys = keys.filter(isKnownKey);
        if (keys.length === knownKeys.length) {
            filteredRows.push(row);
        }
        else {
            const filteredRow = {};
            for (const key of knownKeys) {
                filteredRow[key] = row[key];
            }
            filteredRows.push(filteredRow);
        }
    }
    return filteredRows;
};
exports.filterRowsByColumns = filterRowsByColumns;
