// @airplane/views/dist/client/env.js
var define_import_meta_env_default = { BASE_URL: "/", MODE: "production", DEV: false, PROD: true, SSR: false };
let AIRPLANE_API_HOST = (define_import_meta_env_default == null ? void 0 : define_import_meta_env_default.AIRPLANE_API_HOST) ?? "https://api.airplane.dev";
let AIRPLANE_TOKEN = (define_import_meta_env_default == null ? void 0 : define_import_meta_env_default.AIRPLANE_TOKEN) ?? "";
let AIRPLANE_API_KEY = (define_import_meta_env_default == null ? void 0 : define_import_meta_env_default.AIRPLANE_API_KEY) ?? "";
let AIRPLANE_VIEW_TOKEN = (define_import_meta_env_default == null ? void 0 : define_import_meta_env_default.AIRPLANE_VIEW_TOKEN) ?? "";
let AIRPLANE_ENV_SLUG = (define_import_meta_env_default == null ? void 0 : define_import_meta_env_default.AIRPLANE_ENV_SLUG) ?? "";
let AIRPLANE_TUNNEL_TOKEN = (define_import_meta_env_default == null ? void 0 : define_import_meta_env_default.AIRPLANE_TUNNEL_TOKEN) ?? "";
let AIRPLANE_SANDBOX_TOKEN = (define_import_meta_env_default == null ? void 0 : define_import_meta_env_default.AIRPLANE_SANDBOX_TOKEN) ?? "";
let AIRPLANE_API_HEADERS = (define_import_meta_env_default == null ? void 0 : define_import_meta_env_default.AIRPLANE_API_HEADERS) ?? "{}";
let AIRPLANE_USE_SELF_HOSTED_INPUTS = (define_import_meta_env_default == null ? void 0 : define_import_meta_env_default.AIRPLANE_USE_SELF_HOSTED_INPUTS) === "true";
const setEnvVars = (apiHost, token, apiKey, envSlug, additionalEnvVars) => {
  AIRPLANE_API_HOST = apiHost;
  AIRPLANE_TOKEN = token;
  AIRPLANE_VIEW_TOKEN = additionalEnvVars == null ? void 0 : additionalEnvVars.AIRPLANE_VIEW_TOKEN;
  AIRPLANE_API_KEY = apiKey;
  AIRPLANE_ENV_SLUG = envSlug;
  if (apiHost.includes("stage"))
    ;
  AIRPLANE_TUNNEL_TOKEN = (additionalEnvVars == null ? void 0 : additionalEnvVars.AIRPLANE_TUNNEL_TOKEN) ?? "";
  AIRPLANE_SANDBOX_TOKEN = (additionalEnvVars == null ? void 0 : additionalEnvVars.AIRPLANE_SANDBOX_TOKEN) ?? "";
  AIRPLANE_API_HEADERS = (additionalEnvVars == null ? void 0 : additionalEnvVars.AIRPLANE_API_HEADERS) ?? "{}";
  AIRPLANE_USE_SELF_HOSTED_INPUTS = (additionalEnvVars == null ? void 0 : additionalEnvVars.AIRPLANE_USE_SELF_HOSTED_INPUTS) === "true";
};
const getExecuteOptions = (source, allowCachedMaxAge) => {
  return {
    host: AIRPLANE_API_HOST,
    token: AIRPLANE_TOKEN,
    apiKey: AIRPLANE_TOKEN ? void 0 : AIRPLANE_API_KEY,
    envSlug: AIRPLANE_ENV_SLUG,
    source: `view-${source}`,
    headers: {
      "X-Airplane-View-Token": AIRPLANE_VIEW_TOKEN
    },
    useSelfHostedInputs: AIRPLANE_USE_SELF_HOSTED_INPUTS,
    allowCachedMaxAge
  };
};
export {
  AIRPLANE_API_HEADERS,
  AIRPLANE_API_HOST,
  AIRPLANE_API_KEY,
  AIRPLANE_ENV_SLUG,
  AIRPLANE_SANDBOX_TOKEN,
  AIRPLANE_TOKEN,
  AIRPLANE_TUNNEL_TOKEN,
  AIRPLANE_USE_SELF_HOSTED_INPUTS,
  AIRPLANE_VIEW_TOKEN,
  getExecuteOptions,
  setEnvVars
};
