// @airplane/views/dist/components/multiselect/MultiSelect.js
import { jsx } from "react/jsx-runtime";
import { MultiSelect as MultiSelect$1 } from "@mantine/core";
import { useState, forwardRef } from "react";
import { ComponentErrorBoundary } from "../errorBoundary/ComponentErrorBoundary.js";
import { useSetLatestRunInTaskQuery } from "../errorBoundary/LatestRunDetails.js";
import { useCommonLayoutStyle } from "../layout/useCommonLayoutStyle.js";
import { Loader } from "../loader/Loader.js";
import { displayTaskBackedError } from "../../errors/displayTaskBackedError.js";
import { useRegisterFormInput } from "../../state/components/form/useRegisterFormInput.js";
import { useInput } from "../../state/components/input/useInput.js";
import { useMultiSelectState } from "../../state/components/multiselect/useMultiSelectState.js";
import { useComponentId } from "../../state/components/useId.js";
import { useTaskQuery } from "../../state/tasks/useTaskQuery.js";
const NUMBER_PREFIX = "__airplane_number__";
const defaultProps = {
  searchable: true
};
const MultiSelect = (props) => {
  const [latestRun, setLatestRun] = useState();
  if (doesUseTask(props)) {
    return /* @__PURE__ */ jsx(ComponentErrorBoundary, { componentName: MultiSelect.displayName, latestRun, children: /* @__PURE__ */ jsx(MultiSelectWithTask, { ...props, setLatestRun }) });
  } else {
    return /* @__PURE__ */ jsx(ComponentErrorBoundary, { componentName: MultiSelect.displayName, children: /* @__PURE__ */ jsx(ConnectedMultiSelect, { ...props }) });
  }
};
MultiSelect.displayName = "MultiSelect";
const MultiSelectWithTask = ({
  task,
  outputTransform,
  setLatestRun,
  ...restProps
}) => {
  const fullQuery = useSetLatestRunInTaskQuery(task, setLatestRun);
  const {
    error,
    loading,
    output,
    runID
  } = useTaskQuery(fullQuery);
  const data = output ? outputToData(output, outputTransform) : [];
  if (error) {
    return displayTaskBackedError({
      error,
      taskSlug: fullQuery.slug,
      runID,
      componentName: "MultiSelect"
    });
  } else {
    return /* @__PURE__ */ jsx(ConnectedMultiSelect, { ...restProps, loading, data });
  }
};
const ConnectedMultiSelect = (props) => {
  const id = useComponentId(props.id);
  const {
    state,
    dispatch
  } = useMultiSelectState(id, {
    initialState: {
      disabled: props.disabled ?? props.defaultDisabled,
      value: props.value ?? props.defaultValue
    }
  });
  const propsOnChange = props.onChange;
  const {
    inputProps
  } = useInput({
    ...props,
    onChange: propsOnChange && ((v) => propsOnChange(v.map((vs) => convertMultiSelectStringToOriginalType(vs))))
  }, state, dispatch, (v) => v.map((vs) => convertMultiSelectStringToOriginalType(vs)));
  useRegisterFormInput(id, "multi-select");
  const {
    data,
    validate: _,
    onChange: __,
    defaultDisabled: ___,
    defaultValue: ____,
    error: propsError,
    ...restProps
  } = props;
  const error = propsError || inputProps.error;
  const newData = data.map((item) => {
    if (typeof item === "string") {
      return item;
    } else if (typeof item === "number") {
      return numberToMultiSelectItem(item);
    } else {
      return multiSelectItemToMantine(item);
    }
  });
  return /* @__PURE__ */ jsx(MultiSelectComponent, { data: newData, ...defaultProps, ...inputProps, ...restProps, error });
};
const MultiSelectComponent = /* @__PURE__ */ forwardRef((props, ref) => /* @__PURE__ */ jsx(MultiSelectComponentWithoutRef, { ...props, innerRef: ref }));
MultiSelectComponent.displayName = "MultiSelectComponent";
const MultiSelectComponentWithoutRef = ({
  loading,
  data,
  value,
  defaultValue,
  filter,
  withinPortal,
  innerRef,
  unstyled,
  ItemComponent,
  itemComponent,
  disabled,
  className,
  style,
  width,
  height,
  grow,
  ...restProps
}) => {
  const {
    classes: layoutClasses,
    cx
  } = useCommonLayoutStyle({
    width,
    height,
    grow
  });
  const newProps = {
    data,
    value: value == null ? void 0 : value.map((vs) => convertMultiSelectValueToString(vs)),
    defaultValue: defaultValue == null ? void 0 : defaultValue.map((vs) => convertMultiSelectValueToString(vs)),
    filter: filter ? (value2, selected, item) => {
      return filter(value2, selected, mantineToMultiSelectItem(item));
    } : void 0
  };
  return /* @__PURE__ */ jsx(MultiSelect$1, { withinPortal, ref: innerRef, variant: unstyled ? "unstyled" : void 0, className: cx(layoutClasses.style, className), style, itemComponent: ItemComponent || itemComponent, ...newProps, ...restProps, icon: loading && /* @__PURE__ */ jsx(Loader, { size: "xs", color: "secondary" }), disabled: disabled || loading });
};
function outputToData(output, dataTransform) {
  if (!output) {
    return [];
  }
  if (dataTransform) {
    return dataTransform(output);
  }
  if (Array.isArray(output)) {
    return output;
  }
  const unwrappedOutput = unwrapOutput(output);
  if (unwrappedOutput) {
    return unwrappedOutput;
  }
  return [];
}
function doesUseTask(props) {
  return Boolean(props.task);
}
const unwrapOutput = (data) => {
  if (data && !Array.isArray(data) && typeof data === "object") {
    const keys = Object.keys(data);
    if (keys.length === 1) {
      const value = data[keys[0]];
      if (Array.isArray(value) && value.every((item) => typeof item === "string" || isMultiSelectItem(item))) {
        return value;
      }
    }
  }
  return void 0;
};
const isMultiSelectItem = (item) => !Array.isArray(item) && typeof item === "object" && typeof item.value === "string";
const numberToMultiSelectItem = (value) => {
  return {
    value: convertMultiSelectValueToString(value),
    label: String(value)
  };
};
const convertMultiSelectValueToString = (value) => {
  if (typeof value === "number") {
    return NUMBER_PREFIX + String(value);
  } else {
    return value;
  }
};
const convertMultiSelectStringToOriginalType = (s) => {
  if (s.startsWith(NUMBER_PREFIX)) {
    return Number(s.substring(NUMBER_PREFIX.length));
  } else {
    return s;
  }
};
const multiSelectItemToMantine = (item) => {
  const {
    value,
    ...restFields
  } = item;
  return {
    value: convertMultiSelectValueToString(value),
    ...restFields
  };
};
const mantineToMultiSelectItem = (item) => {
  const {
    value,
    ...restFields
  } = item;
  return {
    value: convertMultiSelectStringToOriginalType(value),
    ...restFields
  };
};
export {
  MultiSelect,
  MultiSelectComponent,
  MultiSelectComponentWithoutRef
};
