// airplane/esm/index.js
import * as api from "./internal/api";
import * as auth from "./internal/auth";
import * as ai from "./internal/builtins/ai";
import * as email from "./internal/builtins/email";
import * as graphql from "./internal/builtins/graphql";
import * as mongodb from "./internal/builtins/mongodb";
import * as rest from "./internal/builtins/rest";
import * as slack from "./internal/builtins/slack";
import * as sql from "./internal/builtins/sql";
import { task, view, workflow } from "./internal/config";
import * as display from "./internal/display";
import { execute, executeBackground, executeRunbook, executeRunbookBackground, } from "./internal/execute";
import { AirplaneFile } from "./internal/file";
import * as file from "./internal/file";
import { appendOutput, setOutput } from "./internal/output";
import { prompt } from "./internal/prompt";
import { sleep } from "./internal/sleep";
export { RunTerminationError, SessionTerminationError, PromptCancelledError, } from "./internal/errors";
// Export the core methods so they can be directly imported:
//
// import { execute } from 'airplane'
export { AirplaneFile, appendOutput, setOutput, execute, executeBackground, executeRunbook, executeRunbookBackground, task, view, workflow, };
// Default export the entire airplane SDK for convenience, e.g.:
//
// import airplane from 'airplane'
// airplane.execute(...)
// airplane.email.message(...)
export default {
    appendOutput,
    setOutput,
    execute,
    executeBackground,
    executeRunbook,
    executeRunbookBackground,
    api,
    ai,
    email,
    graphql,
    mongodb,
    rest,
    slack,
    sql,
    prompt,
    task,
    view,
    workflow,
    sleep,
    display,
    file,
    auth,
};
