// airplane/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.workflow = exports.view = exports.task = exports.setOutput = exports.executeRunbookBackground = exports.executeRunbook = exports.executeBackgroundWithCacheInfo = exports.executeBackground = exports.execute = exports.appendOutput = exports.AirplaneFile = exports.SessionTerminationError = exports.RunTerminationError = exports.PromptCancelledError = void 0;
const api = __importStar(require("./internal/api"));
const auth = __importStar(require("./internal/auth"));
const ai = __importStar(require("./internal/builtins/ai"));
const email = __importStar(require("./internal/builtins/email"));
const graphql = __importStar(require("./internal/builtins/graphql"));
const mongodb = __importStar(require("./internal/builtins/mongodb"));
const rest = __importStar(require("./internal/builtins/rest"));
const slack = __importStar(require("./internal/builtins/slack"));
const sql = __importStar(require("./internal/builtins/sql"));
const config_1 = require("./internal/config");
Object.defineProperty(exports, "task", { enumerable: true, get: function () { return config_1.task; } });
Object.defineProperty(exports, "view", { enumerable: true, get: function () { return config_1.view; } });
Object.defineProperty(exports, "workflow", { enumerable: true, get: function () { return config_1.workflow; } });
const display = __importStar(require("./internal/display"));
const execute_1 = require("./internal/execute");
Object.defineProperty(exports, "execute", { enumerable: true, get: function () { return execute_1.execute; } });
Object.defineProperty(exports, "executeBackground", { enumerable: true, get: function () { return execute_1.executeBackground; } });
Object.defineProperty(exports, "executeBackgroundWithCacheInfo", { enumerable: true, get: function () { return execute_1.executeBackgroundWithCacheInfo; } });
Object.defineProperty(exports, "executeRunbook", { enumerable: true, get: function () { return execute_1.executeRunbook; } });
Object.defineProperty(exports, "executeRunbookBackground", { enumerable: true, get: function () { return execute_1.executeRunbookBackground; } });
const file = __importStar(require("./internal/file"));
const file_1 = require("./internal/file");
Object.defineProperty(exports, "AirplaneFile", { enumerable: true, get: function () { return file_1.AirplaneFile; } });
const output_1 = require("./internal/output");
Object.defineProperty(exports, "appendOutput", { enumerable: true, get: function () { return output_1.appendOutput; } });
Object.defineProperty(exports, "setOutput", { enumerable: true, get: function () { return output_1.setOutput; } });
const prompt_1 = require("./internal/prompt");
const sleep_1 = require("./internal/sleep");
var errors_1 = require("./internal/errors");
Object.defineProperty(exports, "PromptCancelledError", { enumerable: true, get: function () { return errors_1.PromptCancelledError; } });
Object.defineProperty(exports, "RunTerminationError", { enumerable: true, get: function () { return errors_1.RunTerminationError; } });
Object.defineProperty(exports, "SessionTerminationError", { enumerable: true, get: function () { return errors_1.SessionTerminationError; } });
// Default export the entire airplane SDK for convenience, e.g.:
//
// import airplane from 'airplane'
// airplane.execute(...)
// airplane.email.message(...)
exports.default = {
    appendOutput: output_1.appendOutput,
    setOutput: output_1.setOutput,
    execute: execute_1.execute,
    executeBackground: execute_1.executeBackground,
    executeBackgroundWithCacheInfo: execute_1.executeBackgroundWithCacheInfo,
    executeRunbook: execute_1.executeRunbook,
    executeRunbookBackground: execute_1.executeRunbookBackground,
    api,
    ai,
    email,
    graphql,
    mongodb,
    rest,
    slack,
    sql,
    prompt: prompt_1.prompt,
    task: config_1.task,
    view: config_1.view,
    workflow: config_1.workflow,
    sleep: sleep_1.sleep,
    display,
    file,
    auth,
};
